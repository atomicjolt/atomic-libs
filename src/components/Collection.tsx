import { Node } from "react-stately";
import {
  ItemProps as RSItemProps,
  SectionProps as RSSectionProps,
  Item as RSItem,
  Section as RSSection,
} from "react-stately";
import {
  createBranchComponent,
  createLeafComponent,
} from "@react-aria/collections";
import { createContext, useContext } from "react";
import { copyStaticProperties } from "@utils/clone";
import { Key, RenderBaseProps, RenderStyleProps } from "../types";

export interface ElementItemProps<T, R extends object = object>
  extends Omit<RSItemProps<T>, "children">,
    RenderBaseProps<R> {
  id?: Key;
}

interface ItemContextValue {
  render: (
    props: ElementItemProps<any>,
    ref: React.ForwardedRef<HTMLElement>,
    item: Node<any>
  ) => React.ReactElement;
}

export const ItemContext = createContext<ItemContextValue | null>(null);

/** An Item in a Collection component
 * @example
 * <CustomSelect label="Select a fruit">
 *  <Item id="apple">Apple</Item>
 *  <Item id="orange">Orange</Item>
 *  <Item id="banana">Banana</Item>
 * </CustomSelect>
 */
export const Item = createLeafComponent(
  "item",
  (
    props: ElementItemProps<any>,
    ref: React.ForwardedRef<HTMLElement>,
    item: Node<object>
  ) => {
    const { render } = useContext(ItemContext)!;
    return render(props, ref, item);
  }
);

copyStaticProperties(RSItem, Item);

export interface ElementSectionProps<T, R extends object = object>
  extends RSSectionProps<T>,
    RenderStyleProps<R> {
  id?: Key;
}

interface SectionContextValue {
  render: (
    props: ElementSectionProps<any>,
    ref: React.ForwardedRef<HTMLElement>,
    section: Node<any>
  ) => React.ReactElement;
}

export const SectionContext = createContext<SectionContextValue | null>(null);

/** A Section / grouping of items in a collection component
 * @example
 * <ListBox>
 *  <Section title="Fruits">
 *    <Item id="apple">Apple</Item>
 *    <Item id="orange">Orange</Item>
 *    <Item id="banana">Banana</Item>
 *  </Section>
 *  <Section title="Vegetables">
 *    <Item id="carrot">Carrot</Item>
 *    <Item id="potato">Potato</Item>
 *    <Item id="onion">Onion</Item>
 *  </Section>
 * </ListBox>
 */
export const Section = createBranchComponent(
  "section",
  (
    props: ElementSectionProps<any>,
    ref: React.ForwardedRef<HTMLElement>,
    section: Node<object>
  ) => {
    const { render } = useContext(SectionContext)!;
    return render(props, ref, section);
  }
);

copyStaticProperties(RSSection, Section);

export type {
  ElementItemProps as ItemProps,
  ElementSectionProps as SectionProps,
};
