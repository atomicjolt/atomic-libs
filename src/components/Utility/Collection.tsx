import {
  Item as ReactStatelyItem,
  Section as ReactStatelySection,
  ItemProps,
  SectionProps,
} from "react-stately";

// We re-define the react-stately Item and Section components
// Because for some reason storybook doesn't like to pickup the
// correct names for the reac-stately components.

/** An Item in a Collection component
 * @example
 * <CustomSelect label="Select a fruit">
 *  <Item key="apple">Apple</Item>
 *  <Item key="orange">Orange</Item>
 *  <Item key="banana">Banana</Item>
 * </CustomSelect>
 */
export function Item<T>(props: ItemProps<T>) {
  return null;
}

// React Stately binds some static properties to the Item component that we need to copy over
for (const key in ReactStatelyItem) {
  // @ts-ignore
  Item[key] = ReactStatelyItem[key];
}

/** A Section / grouping of items in a collection component
 * @example
 * <ListBox>
 *  <Section title="Fruits">
 *    <Item key="apple">Apple</Item>
 *    <Item key="orange">Orange</Item>
 *    <Item key="banana">Banana</Item>
 *  </Section>
 *  <Section title="Vegetables">
 *    <Item key="carrot">Carrot</Item>
 *    <Item key="potato">Potato</Item>
 *    <Item key="onion">Onion</Item>
 *  </Section>
 * </ListBox>
 */
export function Section<T>(props: SectionProps<T>) {
  return null;
}

for (const key in ReactStatelySection) {
  // @ts-ignore
  Section[key] = ReactStatelySection[key];
}
