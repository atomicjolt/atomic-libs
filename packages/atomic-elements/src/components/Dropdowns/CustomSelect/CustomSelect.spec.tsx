import { describe, test, expect } from "vitest";
import { render } from "@testing-library/react";
import { CustomSelect } from ".";
import { Item } from "../../Collection";

describe("matches snapshots", () => {
  const shared = {
    label: "Custom dropdown label",
    value: "opt1",
    onChange: () => {},
  };

  test("matches default variant", () => {
    const result = render(
      <CustomSelect {...shared}>
        <Item key="opt1">Item 1</Item>
        <Item key="opt2">Item 2</Item>
      </CustomSelect>
    );

    expect(result.asFragment()).toMatchSnapshot();
  });

  // test("matches multi-select enabled", () => {
  //   const result = render(
  //     <CustomSelect {...shared} value={["opt1", "opt2"]}>
  //       <Item key="opt1">Item 1</Item>
  //       <Item key="opt2">Item 2</Item>
  //     </CustomSelect>
  //   );

  //   expect(result.asFragment()).toMatchSnapshot();
  // });

  // TODO: Searching has not been reimplemented
  // test("matches search enabled", () => {
  //   const result = render(
  //     <CustomSelect {...shared} searchable>
  //       <Item key="opt1">Item 1</Item>
  //       <Item key="opt2">Item 2</Item>
  //     </CustomSelect>
  //   );

  //   expect(result.asFragment()).toMatchSnapshot();
  // });

  test("matches floating variant", () => {
    const result = render(
      <CustomSelect variant="floating" {...shared}>
        <Item key="opt1">Item 1</Item>
        <Item key="opt2">Item 2</Item>
      </CustomSelect>
    );

    expect(result.asFragment()).toMatchSnapshot();
  });
});
