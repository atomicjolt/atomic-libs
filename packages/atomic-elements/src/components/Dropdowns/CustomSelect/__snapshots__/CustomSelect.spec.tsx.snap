// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshots > matches default variant 1`] = `
<DocumentFragment>
  <template
    data-react-aria-hidden="true"
  />
  <div
    class="sc-gpHbIA bhSUCw aje-select sc-gQJZgv fvvEjB aje-select aje-select--default"
  >
    <div
      aria-hidden="true"
      data-a11y-ignore="aria-hidden-focus"
      data-react-aria-prevent-focus="true"
      data-testid="hidden-select-container"
      style="border: 0px; clip-path: inset(50%); height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
    >
      <label>
        <select
          tabindex="-1"
        >
          <option />
          <option
            value="react-aria-1"
          >
            Item 1
          </option>
          <option
            value="react-aria-2"
          >
            Item 2
          </option>
        </select>
      </label>
    </div>
    <label
      class="sc-dVBluf eSKkgA"
      id="react-aria-:r3:"
    >
      Custom dropdown label
    </label>
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-labelledby="react-aria-:r7: react-aria-:r3:"
      class="sc-eiwqOE JUNDn aje-btn aje-btn--dropdown is-medium"
      id="react-aria-:r2:"
      tabindex="0"
      type="button"
    >
      <span
        class="sc-hUAZgl copoPc"
        id="react-aria-:r7:"
      >
        Select an option
      </span>
      <i
        aria-hidden="true"
        class="sc-jBIHhB dShSRo aje-icon material-icons is-medium"
        style="margin-left: auto;"
      >
        arrow_drop_down
      </i>
    </button>
  </div>
</DocumentFragment>
`;

exports[`matches snapshots > matches floating variant 1`] = `
<DocumentFragment>
  <template
    data-react-aria-hidden="true"
  />
  <div
    class="sc-gpHbIA bhSUCw aje-select sc-gQJZgv fvvEjB aje-select aje-select--floating aje-select--floating"
  >
    <div
      aria-hidden="true"
      data-a11y-ignore="aria-hidden-focus"
      data-react-aria-prevent-focus="true"
      data-testid="hidden-select-container"
      style="border: 0px; clip-path: inset(50%); height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
    >
      <label>
        <select
          tabindex="-1"
        >
          <option />
          <option
            value="react-aria-1"
          >
            Item 1
          </option>
          <option
            value="react-aria-2"
          >
            Item 2
          </option>
        </select>
      </label>
    </div>
    <div
      class="sc-flQbIK llRNxW"
    >
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="react-aria-:rf: react-aria-:rb:"
        class="sc-eiwqOE JUNDn aje-btn aje-btn--dropdown is-medium"
        id="react-aria-:ra:"
        tabindex="0"
        type="button"
      >
        <span
          class="sc-hUAZgl copoPc"
          id="react-aria-:rf:"
        />
        <i
          aria-hidden="true"
          class="sc-jBIHhB dShSRo aje-icon material-icons is-medium"
          style="margin-left: auto;"
        >
          arrow_drop_down
        </i>
      </button>
      <label
        class="sc-dVBluf eSKkgA"
        id="react-aria-:rb:"
      >
        Custom dropdown label
      </label>
    </div>
  </div>
</DocumentFragment>
`;
